File: frontend/src/api/portfolio.ts
import { ApiClient } from './request'

export interface PortfolioHolding {
  id: string
  stock_code: string
  stock_name: string
  market: string
  quantity: number
  avg_price: number
  buy_date: string
  notes: string
  created_at?: string
  updated_at?: string
}

export interface AddHoldingRequest {
  stock_code: string
  stock_name: string
  market?: string
  quantity: number
  avg_price: number
  buy_date: string
  notes?: string
}

export interface UpdateHoldingRequest {
  quantity?: number
  avg_price?: number
  buy_date?: string
  notes?: string
}

export const portfolioApi = {
  /**
   * 获取持仓列表
   */
  list(): Promise<{ success: boolean; data: PortfolioHolding[]; message?: string }> {
    return ApiClient.get('/api/portfolio/')
  },

  /**
   * 添加持仓
   */
  add(request: AddHoldingRequest): Promise<{ success: boolean; data: PortfolioHolding; message?: string }> {
    return ApiClient.post('/api/portfolio/', request)
  },

  /**
   * 更新持仓
   */
  update(
    holdingId: string,
    request: UpdateHoldingRequest
  ): Promise<{ success: boolean; data: PortfolioHolding; message?: string }> {
    return ApiClient.put(`/api/portfolio/${holdingId}`, request)
  },

  /**
   * 删除持仓
   */
  remove(holdingId: string): Promise<{ success: boolean; data: { id: string }; message?: string }> {
    return ApiClient.delete(`/api/portfolio/${holdingId}`)
  },

  /**
   * 获取单条持仓
   */
  get(holdingId: string): Promise<{ success: boolean; data: PortfolioHolding; message?: string }> {
    return ApiClient.get(`/api/portfolio/${holdingId}`)
  },
}

// ==================== 基金持仓 ====================

export interface FundHolding {
  id: string
  fund_code: string
  fund_name: string
  fund_type: string
  quantity: number
  avg_nav: number
  buy_date: string
  notes: string
  created_at?: string
  updated_at?: string
}

export interface AddFundHoldingRequest {
  fund_code: string
  fund_name: string
  fund_type?: string
  quantity: number
  avg_nav: number
  buy_date: string
  notes?: string
}

export interface UpdateFundHoldingRequest {
  quantity?: number
  avg_nav?: number
  buy_date?: string
  notes?: string
}

// ==================== 现金管理 ====================

export interface CashItem {
  id: string
  currency: string
  amount: number
  notes: string
  created_at?: string
  updated_at?: string
}

export interface AddCashRequest {
  currency: string
  amount: number
  notes?: string
}

export const cashApi = {
  list(): Promise<{ success: boolean; data: CashItem[]; message?: string }> {
    return ApiClient.get('/api/portfolio/cash/')
  },
  add(request: AddCashRequest): Promise<{ success: boolean; data: CashItem; message?: string }> {
    return ApiClient.post('/api/portfolio/cash/', request)
  },
  update(cashId: string, request: Partial<AddCashRequest>): Promise<{ success: boolean; data: CashItem; message?: string }> {
    return ApiClient.put(`/api/portfolio/cash/${cashId}`, request)
  },
  remove(cashId: string): Promise<{ success: boolean; data: { id: string }; message?: string }> {
    return ApiClient.delete(`/api/portfolio/cash/${cashId}`)
  }
}

export const fundPortfolioApi = {
  /**
   * 获取基金持仓列表
   */
  list(): Promise<{ success: boolean; data: FundHolding[]; message?: string }> {
    return ApiClient.get('/api/portfolio/fund/')
  },

  /**
   * 添加基金持仓
   */
  add(request: AddFundHoldingRequest): Promise<{ success: boolean; data: FundHolding; message?: string }> {
    return ApiClient.post('/api/portfolio/fund/', request)
  },

  /**
   * 更新基金持仓
   */
  update(
    holdingId: string,
    request: UpdateFundHoldingRequest
  ): Promise<{ success: boolean; data: FundHolding; message?: string }> {
    return ApiClient.put(`/api/portfolio/fund/${holdingId}`, request)
  },

  /**
   * 删除基金持仓
   */
  remove(holdingId: string): Promise<{ success: boolean; data: { id: string }; message?: string }> {
    return ApiClient.delete(`/api/portfolio/fund/${holdingId}`)
  },
}

File: app/models/portfolio.py
"""
投资组合持仓数据模型
"""

from datetime import datetime
from typing import Optional
from pydantic import BaseModel, Field
from app.utils.timezone import now_tz


class PortfolioHolding(BaseModel):
    """持仓记录模型"""
    id: Optional[str] = Field(None, description="持仓记录ID")
    stock_code: str = Field(..., description="股票代码")
    stock_name: str = Field(..., description="股票名称")
    market: str = Field(default="A股", description="市场类型")
    quantity: int = Field(..., ge=1, description="持有数量")
    avg_price: float = Field(..., ge=0, description="买进均价")
    buy_date: str = Field(..., description="买进时间 (YYYY-MM-DD)")
    notes: str = Field(default="", description="备注")
    created_at: Optional[datetime] = Field(default_factory=now_tz, description="创建时间")
    updated_at: Optional[datetime] = Field(default_factory=now_tz, description="更新时间")


class AddHoldingRequest(BaseModel):
    """添加持仓请求"""
    stock_code: str = Field(..., description="股票代码")
    stock_name: str = Field(..., description="股票名称")
    market: str = Field(default="A股", description="市场类型")
    quantity: int = Field(..., ge=1, description="持有数量")
    avg_price: float = Field(..., ge=0, description="买进均价")
    buy_date: str = Field(..., description="买进时间 (YYYY-MM-DD)")
    notes: str = Field(default="", description="备注")


class UpdateHoldingRequest(BaseModel):
    """更新持仓请求"""
    quantity: Optional[int] = Field(None, ge=1, description="持有数量")
    avg_price: Optional[float] = Field(None, ge=0, description="买进均价")
    buy_date: Optional[str] = Field(None, description="买进时间 (YYYY-MM-DD)")
    notes: Optional[str] = Field(None, description="备注")


class HoldingResponse(BaseModel):
    """持仓响应"""
    id: str
    stock_code: str
    stock_name: str
    market: str
    quantity: int
    avg_price: float
    buy_date: str
    notes: str
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


# ==================== 基金持仓 ====================

class FundHolding(BaseModel):
    """基金持仓记录模型"""
    id: Optional[str] = Field(None, description="持仓记录ID")
    fund_code: str = Field(..., description="基金代码")
    fund_name: str = Field(..., description="基金名称")
    fund_type: str = Field(default="混合型", description="基金类型")
    quantity: float = Field(..., ge=0, description="持有份额")
    avg_nav: float = Field(..., ge=0, description="买入净值")
    buy_date: str = Field(..., description="买进时间 (YYYY-MM-DD)")
    notes: str = Field(default="", description="备注")
    created_at: Optional[datetime] = Field(default_factory=now_tz, description="创建时间")
    updated_at: Optional[datetime] = Field(default_factory=now_tz, description="更新时间")


class AddFundHoldingRequest(BaseModel):
    """添加基金持仓请求"""
    fund_code: str = Field(..., description="基金代码")
    fund_name: str = Field(..., description="基金名称")
    fund_type: str = Field(default="混合型", description="基金类型")
    quantity: float = Field(..., ge=0, description="持有份额")
    avg_nav: float = Field(..., ge=0, description="买入净值")
    buy_date: str = Field(..., description="买进时间 (YYYY-MM-DD)")
    notes: str = Field(default="", description="备注")


class UpdateFundHoldingRequest(BaseModel):
    """更新基金持仓请求"""
    quantity: Optional[float] = Field(None, ge=0, description="持有份额")
    avg_nav: Optional[float] = Field(None, ge=0, description="买入净值")
    buy_date: Optional[str] = Field(None, description="买进时间 (YYYY-MM-DD)")
    notes: Optional[str] = Field(None, description="备注")


class FundHoldingResponse(BaseModel):
    """基金持仓响应"""
    id: str
    fund_code: str
    fund_name: str
    fund_type: str
    quantity: float
    avg_nav: float
    buy_date: str
    notes: str
    created_at: Optional[str] = None
    updated_at: Optional[str] = None
